data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/producer/KafkaSinkProducer.java

```diff
@@ -0,0 +1,196 @@
+        Object dataForDlq = event.toJsonString();
+        LOG.info("Producing record " + dataForDlq);
+        try {
+            final String serdeFormat = kafkaSinkConfig.getSerdeFormat();
+            if (MessageFormat.JSON.toString().equalsIgnoreCase(serdeFormat)) {
+                publishJsonMessage(record, topic, key, dataForDlq);
+            } else if (MessageFormat.AVRO.toString().equalsIgnoreCase(serdeFormat)) {
+                publishAvroMessage(record, topic, key, dataForDlq);
+            } else {
+                publishPlaintextMessage(record, topic, key, dataForDlq);
+            }
+        } catch (Exception e) {
+            releaseEventHandles(false);
+        }
+
+    }
+
+    private Event getEvent(Record<Event> record) {
+        Event event = record.getData();
+        try {
+            event = addTagsToEvent(event, tagTargetKey);
+        } catch (JsonProcessingException e) {
+            LOG.error("error occured while processing tag target key");
+        }
+        return event;
+    }
+
+    private void publishPlaintextMessage(Record<Event> record, TopicConfig topic, String key, Object dataForDlq) {
+        producer.send(new ProducerRecord(topic.getName(), key, record.getData().toJsonString()), callBack(dataForDlq));
+    }
+
+    private void publishAvroMessage(Record<Event> record, TopicConfig topic, String key, Object dataForDlq) throws RestClientException, IOException {
+        final String valueToParse = schemaRegistryClient.
+                getLatestSchemaMetadata(topic.getName() + "-value").getSchema();
+        final Schema schema = new Schema.Parser().parse(valueToParse);
+        final GenericRecord genericRecord = getGenericRecord(record.getData(), schema);
+        producer.send(new ProducerRecord(topic.getName(), key, genericRecord), callBack(dataForDlq));
+    }
+
+    private void publishJsonMessage(Record<Event> record, TopicConfig topic, String key, Object dataForDlq) throws IOException, RestClientException, ProcessingException {
+        final JsonNode dataNode = new ObjectMapper().convertValue(record.getData().toJsonString(), JsonNode.class);
+        if (validateJson(topic.getName(), dataForDlq)) {
+            producer.send(new ProducerRecord(topic.getName(), key, dataNode), callBack(dataForDlq));
+        } else {
+            dlqSink.perform(dataForDlq, new RuntimeException("Invalid Json"));
+        }
+    }
+
+    private Boolean validateJson(final String topicName, Object dataForDlq) throws IOException, RestClientException, ProcessingException {
+        if (schemaRegistryClient != null) {
+            final String schemaJson = schemaRegistryClient.
+                    getLatestSchemaMetadata(topicName + "-value").getSchema();
+            return validateSchema(dataForDlq.toString(), schemaJson);
+        } else {
+            return true;
+        }
+    }
+
+    private boolean validateSchema(final String jsonData, final String schemaJson) throws IOException, ProcessingException {
```

Comment:
does this need to happen for every produce record?

Reply:
yes , this is implemented as per the requirement.